ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/CustodyGroupCountManagerImpl.java

```diff
@@ -93,29 +94,69 @@ public void onSlot(final UInt64 slot) {
       return;
     }
 
-    if (!updateEpoch(spec.computeEpochAtSlot(slot))) {
+    final UInt64 currentEpoch = spec.computeEpochAtSlot(slot);
+
+    final Map<UInt64, PreparedProposerInfo> preparedValidators =
+        proposersDataManager.getPreparedProposerInfo();
+
+    if (detectGenesisInitialization(currentEpoch, preparedValidators)) {
+      updateEpoch(currentEpoch);
       return;
     }
 
-    final Map<UInt64, PreparedProposerInfo> preparedProposerInfo =
-        proposersDataManager.getPreparedProposerInfo();
-    if (preparedProposerInfo.isEmpty()) {
+    if (!updateEpoch(currentEpoch)) {
+      return;
+    }
+
+    if (preparedValidators.isEmpty()) {
       updateCustodyGroupCount(initCustodyGroupCount);
       return;
     }
 
+    computeAndUpdateCustodyGroupCount(preparedValidators);
+  }
+
+  private boolean detectGenesisInitialization(
+      final UInt64 currentEpoch, final Map<UInt64, PreparedProposerInfo> preparedValidators) {
+    if (!currentEpoch.isZero()) {
+      return false;
+    }
+
+    if (preparedValidators.isEmpty()) {
+      return false;
+    }
+
+    if (genesisInitialized) {
+      return false;
+    }
+
+    genesisInitialized = true;
+
+    LOG.info("Validators at genesis epoch detected, initializing custody group count.");
+    computeAndUpdateCustodyGroupCount(preparedValidators);
+    return true;
+  }
+
+  private void computeAndUpdateCustodyGroupCount(
+      final Map<UInt64, PreparedProposerInfo> preparedProposerInfo) {
     combinedChainDataClient
         .getBestFinalizedState()
         .thenAccept(
             maybeState -> {
-              if (maybeState.isPresent()) {
-                final int custodyGroupCountUpdated =
-                    miscHelpersFulu
-                        .getValidatorsCustodyRequirement(
-                            maybeState.get(), preparedProposerInfo.keySet())
-                        .max(initCustodyGroupCount)
-                        .intValue();
-                updateCustodyGroupCount(custodyGroupCountUpdated);
+              if (maybeState.isEmpty()) {
+                return;
+              }
+
+              final int custodyGroupCountUpdated =
+                  miscHelpersFulu
+                      .getValidatorsCustodyRequirement(
+                          maybeState.get(), preparedProposerInfo.keySet())
+                      .max(initCustodyGroupCount)
+                      .intValue();
+              updateCustodyGroupCount(custodyGroupCountUpdated);
+              if (maybeState.get().getSlot().isZero()) {
```

Comment:
I don't like it being here. This method and part of code designed to handle custody group count, not synced. How about return SafeFuture<Void> from this method and act where we called it.